### ISSUE TITLE
Add automatic document OCR with Baidu Unlimited-OCR

### ISSUE THREAD

**Issue by MartinVerges** [croit/llm-gateway#15]
## Summary

Add first-class OCR support for uploaded documents using [`baidu/Unlimited-OCR`](https://huggingface.co/baidu/Unlimited-OCR). Uploaded scans, images, and layout-sensitive documents should be converted into high-fidelity OCR text and made available to the selected chat model together with the original attachment reference.

The implementation must reuse the existing attachment storage, PDF rendering, sandbox-runner, upstream routing, tool loop, and usage accounting instead of introducing a parallel upload pipeline.

## Motivation

The gateway currently stores uploads in S3 and exposes them through `fetch_attachment`. Text files and born-digital PDFs work well, but scanned PDFs and image-heavy documents depend on the chat model deciding to fetch and inspect images itself. This is slow, inconsistent, and often loses table/layout information.

Unlimited-OCR is designed for long-horizon document parsing and supports multi-page image input. Its output should become a cached document derivative that gives downstream LLMs more reliable information without requiring every chat model to be vision-capable.

## Proposed flow

```text
upload -> S3 attachment
       -> classify file
       -> extract native text when appropriate
       -> render PDF/Office pages when OCR is needed
       -> dedicated Unlimited-OCR upstream
       -> cached Markdown/structured OCR derivative
       -> chat context receives OCR text + original attachment reference
```

OCR content must be marked as untrusted document data. Instructions found inside an uploaded document must not be treated as gateway or system instructions.

### Required

- Add an internal OCR client using the existing HTTP/upstream infrastructure.
- Support image uploads and scanned PDFs first.
- Reuse `server::pdf::render_pages` or the sandbox-runner for PDF-to-PNG conversion.
- Preserve page order and page numbers.
- Store OCR output as an attachment derivative without modifying the original upload.
- Cache OCR results by document hash, model revision, prompt version, and rendering settings.
- Add an explicit `auto`/`ocr` path to `fetch_attachment` or the internal attachment pipeline.
- Inject completed OCR content into the chat request in a clearly delimited, untrusted-data block.
- Add size, page, pixel, output, timeout, and concurrency limits.
- Surface queued/running/completed/failed status to the chat UI.
- Add usage accounting for OCR work and meaningful operator errors.
- Add offline tests with `wiremock`, including multi-page ordering, cache hits, upstream errors, and malicious OCR text.

### Follow-up

- Add a dedicated `ocr` upstream pool kind and admin UI support if OCR backends must be isolated from chat pools.
- Add background OCR jobs after upload.
- Add structured page/table/bounding-box output.
- Add OCR-aware RAG indexing and page-level retrieval.
- Add Office layout OCR through LibreOffice-to-PDF conversion.

## Technical notes

- Unlimited-OCR expects image input; PDFs must be rasterized first.
- Multi-page parsing uses the model's base image mode.
- The model card documents `temperature=0`, `skip_special_tokens=false`, and repetition-control parameters. The exact vLLM recipe and image must be pinned and tested rather than guessed.
- Prefer the existing `reqwest`, `pdfium-render`, and sandbox dependencies. Any new Cargo dependency requires an entry in `docs/dependencies.md`.
- The model is MIT licensed but uses custom model code; pin the model revision and deployment image.

## Acceptance criteria

- A scanned PDF uploaded in the chat produces usable OCR Markdown without a model tool call.
- A normal text PDF does not invoke OCR in `auto` mode unless explicitly requested.
- OCR output is reused on subsequent turns and after a gateway restart.
- The original file remains downloadable and unchanged.
- A 10-page document preserves page order and reports whether all pages were processed.
- OCR upstream failures produce an actionable user-visible error and do not lose the upload.
- OCR text is never inserted as a system message.
- Existing attachment, chat, tool-loop, `/v1` API, and sandbox tests continue to pass.

## Non-goals

- Replacing native extraction for text-centric files.
- Making Unlimited-OCR available as a general-purpose chat model.
- Sending private S3 URLs directly to external providers.
- Silently OCR-ing every large document without limits or cost visibility.

## Implementation plan

1. Extract/reuse a bounded PDF-to-PNG rendering service.
2. Implement the OCR request/response adapter and a fake upstream test.
3. Add derivative persistence and cache lookup.
4. Integrate `auto`/`ocr` attachment resolution into the chat path.
5. Add UI status and operator configuration.
6. Add documentation, metrics, limits, and production deployment guidance.

**Comment by MartinVerges** [comment]
Initial implementation started in the working tree:\n\n- Added bounded `pdf::RenderSettings` with page, width, and height limits while preserving the existing `render_pages` defaults.\n- Added a reusable `server::ocr` adapter for OpenAI-compatible Unlimited-OCR requests, including ordered PNG `image_url` parts, multi-page mode selection, repetition-control parameters, response parsing, and wiremock coverage.\n- The adapter is intentionally not wired into automatic attachment processing yet; routing currently uses the chat pool as an interim implementation. The next step is derivative persistence and a dedicated OCR pool/configuration before enabling automatic OCR.

**Comment by MartinVerges** [comment]
Verification correction: the official `infer.py --pdf` command does accept a PDF, but it is a convenience wrapper that uses PyMuPDF to rasterize PDF pages and then calls multi-image inference. The model/inference API itself receives images. The vLLM recipe also confirms image content plus the required `vllm_xargs` fields; the earlier adapter used SGLang-style fields, so that has been corrected in the working tree. The vLLM server still needs `--logits_processors vllm.model_executor.models.unlimited_ocr:NGramPerReqLogitsProcessor`.

**Comment by MartinVerges** [comment]
The OCR adapter has now been cleaned up for the PDF-aware sidecar plan. It no longer renders PDFs or constructs a vLLM image request. It sends the original document as multipart to `<ocr-backend>/ocr`; the sidecar owns `infer.py --pdf`/PyMuPDF, Unlimited-OCR inference, vLLM logits configuration, and grounding-token cleanup. The gateway-side contract and tests are documented in `docs/ocr.md`. The sidecar implementation and automatic chat attachment integration remain the next work items.

**Comment by MartinVerges** [comment]
Implemented the next slice in commit `bcdf7a5`: added the PDF-aware `deploy/ocr-sidecar` (PyMuPDF + internal `/ocr` multipart contract + vLLM call), Compose `ocr` profile, `[chat.ocr]` configuration, and automatic current-turn PDF/image enrichment. The gateway checks for an enabled, routable and healthy `ocr` model before fetching any attachment; without that backend the feature is completely inactive and no OCR tool/model is exposed. Full suite: 1461 passed, 1 skipped; lint passed.

**Comment by MartinVerges** [comment]
Completed in `d930ffb` (merged to local `main` as `1252860`). The sidecar work in `bcdf7a5` left the required scope half-open — OCR ran on every PDF and image of every turn, threw the result away at the end of it, and reported failures only to the log. That is now closed:

**Cache** — migration `0054_ocr_derivatives.sql` + `db::ocr_derivatives`: one row per document hash + model + prompt version + settings fingerprint (prompt, `max_tokens`, `ngram_window`, `max_pages`, `dpi`, `max_output_chars`). Operational knobs stay out of the key — they don't change the text. A `failed` row is kept for the operator but reads as a cache miss, so a transient backend failure retries instead of poisoning the document. Lifecycle `queued → running → completed | failed` is persisted, so it survives a restart.

**`auto` mode** — `pdf_needs_ocr`: a PDF goes to OCR only when fewer than half its pages carry a text layer worth trusting (character-counted, so it holds for every script). A born-digital PDF no longer costs GPU time.

**`fetch_attachment`** — gains `mode="ocr"` (also valid on an image; the only way a text-only model reads one) and `mode="auto"` (text-or-OCR decided server-side for a PDF). Both degrade to the text tier and name the alternative when no OCR backend is configured.

**Status in the chat UI** — each document gets a `document_ocr` row in the turn's activity list, which reuses the existing tool-call rendering for queued/running/completed/failed with a reload-surviving error and a detail panel (pages, `truncated`, `cached`, chars), plus info banners when queued behind the concurrency gate or on failure. The model never sees that row — it is not in the upstream message list and no tool by that name exists — so automatic OCR still needs no model tool call. A failure says the upload is unchanged and still readable, and never fails the turn.

**Metering** — `UsageKind::Ocr`, one `usage_events` row per upstream call including failures, sidecar token counts, pages in `input_units`. Cache hits are free and unrecorded.

**Limits** — `[chat.ocr]`: `max_bytes`, `max_pages`, `dpi`, `max_output_chars`, `timeout_secs`, `max_concurrency` (a gateway-wide semaphore, so it actually bounds anything), `auto_min_text_chars_per_page`; plus hard page/DPI ceilings in the sidecar.

**Sidecar** — one inference call per page by default, so page numbers survive, a single bad page doesn't lose the document, and `pages_processed` is honest; `OCR_MULTI_IMAGE=1` restores the single multi-image call. The gateway sorts page blocks by number before assembling and prefixes each with `--- page N ---`. 400 for a bad request, 502 when the model backend failed.

**Tests** — ~35 new: page ordering from deliberately out-of-order input, cache hit (one upstream call for two recognitions) and miss on changed bytes/settings, upstream failure surfacing and retrying, malicious OCR text staying fenced inside the untrusted-data block in the *user* message, byte and output-char limits, scan detection, the activity-row contract. Full suite on merged `main`: 1666 passed, 1 skipped; `clippy -D warnings` and `fmt` clean.

Docs: `docs/ocr.md` rewritten (cache semantics, status mapping, limits table, both response shapes); README, `gateway.example.toml`, `deploy/README.md`, `compose.example.yml`, `tools-inventory.md`, `file-conversions.md` updated.

Two deliberate scope notes:

- *"Reused on subsequent turns"* is a cache hit, not re-injection. Automatic injection stays current-turn only — re-injecting a 40-page OCR blob every turn is exactly the context bloat the attachment-stub design exists to avoid. On later turns the model gets the text through `fetch_attachment` at zero upstream cost.
- Not yet validated end-to-end against a live sidecar or GPU: there is no S3 mock in the integration harness, so the automatic path's S3 fetch + activity-row loop is covered at unit level only. The per-page `/ocr` contract and the DPI/page limits want one real run on the host before this is trusted in production.

The follow-up list in this issue (background OCR jobs after upload, structured table/bounding-box output, OCR-aware RAG indexing, Office layout OCR via LibreOffice) is untouched and still open — worth a separate issue if wanted. The dedicated `ocr` pool kind and admin UI support it also listed as follow-up already exist.
